clash-lib/src/Clash/Normalize/Transformations.hs

```diff
@@ -2391,3 +2393,101 @@ separateArguments (TransformContext is0 _) e@(collectArgsTicks -> (Var g, args,
         return [(ty,arg)]
 
 separateArguments _ e = return e
+
+-- | Remove all undefined alternatives from case expressions, replacing them
+-- with the value of another defined alternative. If there is one defined
+-- alternative, the entire expression is replaced with that alternative. If
+-- there are no defined alternatives, the entire expression is replaced with
+-- a call to 'errorX'.
+--
+-- e.g. It converts
+--
+--     case x of
+--       D1 a -> f a
+--       D2   -> undefined
+--       D3   -> undefined
+--
+-- to
+--
+--     let subj = x
+--         a    = case subj of
+--                  D1 a -> field0
+--      in f a
+--
+-- where fieldN is an internal variable referring to the nth argument of a
+-- data constructor.
+--
+xOptimize :: HasCallStack => NormRewrite
+xOptimize (TransformContext is0 _) e@(Case subj ty alts) = do
+  defPart <- partitionM (isPrimError . snd) alts
+
+  case defPart of
+    ([], _)    -> return e
+    (_, [])    -> changed (Prim "Clash.XException.errorX" (PrimInfo ty WorkNever))
+    (_, [alt]) -> xOptimizeSingle is0 subj alt
+    (_, defs)  -> xOptimizeMany is0 subj ty defs
+
+xOptimize _ e = return e
+
+-- Return an expression equivalent to the alternative given. When only one
+-- alternative is defined the result of this function is used to replace the
+-- case expression.
+--
+xOptimizeSingle :: InScopeSet -> Term -> Alt -> NormalizeSession Term
+xOptimizeSingle is subj (DataPat dc tvs vars, expr) = do
+  tcm    <- Lens.view tcCache
+  subjId <- mkInternalVar is "subj" (termType tcm subj)
+
+  let fieldTys = fmap varType vars
+  lets <- Monad.zipWithM (mkFieldSelector is subjId dc tvs fieldTys) vars [0..]
+
+  changed (Letrec ((subjId, subj) : lets) expr)
+
+xOptimizeSingle _ _ (_, expr) = changed expr
+
+-- Given a list of alternatives which are defined, create a new case
+-- expression which only ever returns a defined value.
+--
+xOptimizeMany :: InScopeSet -> Term -> Type -> [Alt] -> NormalizeSession Term
+xOptimizeMany is subj ty defs
+  | isAnyDefault defs = changed (Case subj ty defs)
+  | otherwise = do
+      newAlt <- xOptimizeSingle is subj (head defs)
+      changed (Case subj ty $ tail defs <> [(DefaultPat, newAlt)])
```

Comment:
Could you eliminate the use of `head`/`tail` by pattern matching? Then add a branch with a (more or less) clear `error` and a `HasCallStack` constraint. (I know this will never get triggered in the current codebase, but refactors happen..)

Reason: callstacks are often disabled in production code, so you'd just get an annoying `*** Exception: Prelude.head: empty list`. If this is triggered on a clients product, it would mean recompiling everything again..